### PR TITLE
SRA Downloader

### PR DESCRIPTION
**Pull request by kurtwheeler**
This PR adds a surveyor and a Downloader for SRA data.

The SRA surveyor works by taking two accessions, a start and an end. When it comes time to download EVERYTHING, I will scrape these accessions from a file the DDBJ makes public which lists all the accesssions. Therefore being able to specify a range and let it do it's work will make it nice and easy to incrementally increase the amount I run at a time. The surveyor then scrapes data from ENA's XML ftp server. I tried to err on the side of scraping too much metadata rather than too little, but there are a few fields I chose not to scrape because they were redundant or just looked useless out of context.

The downloader doesn't do anything too special. It mostly operates the same as the Array Express downloader except it downloads two files for every Batch.

### CODE REVIEW

File: foreman/data_refinery_foreman/surveyor/test_sra_xml.py [hunk 1]
```diff
@@ -0,0 +1,388 @@
+EXPERIMENT_XML = """<?xml version="1.0" encoding="UTF-8"?>
```
Comment (dongbohu) [inline foreman/data_refinery_foreman/surveyor/test_sra_xml.py]: This file is huge. Do I need to check it line by line?

File: foreman/data_refinery_foreman/surveyor/sra.py [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+)
+from data_refinery_foreman.surveyor.external_source import ExternalSourceSurveyor
+from data_refinery_common.job_lookup import ProcessorPipeline, Downloaders
+from data_refinery_common.logging import get_and_configure_logger
+
+
+logger = get_and_configure_logger(__name__)
+
+
+ENA_METADATA_URL_TEMPLATE = "https://www.ebi.ac.uk/ena/data/view/{}&display=xml"
```
[the remark this answers is not in the document]
Comment (kurtwheeler) [reply foreman/data_refinery_foreman/surveyor/sra.py]: [European Nucleotide Archive](https://www.ebi.ac.uk/ena)

File: foreman/data_refinery_foreman/surveyor/sra.py [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+        return Downloaders.SRA.value
+
+    def determine_pipeline(self,
+                           batch: Batch,
+                           key_values: List[BatchKeyValue] = []):
+        return ProcessorPipeline.SALMON
+
+    @staticmethod
+    def gather_submission_metadata(metadata: Dict) -> None:
+        response = requests.get(ENA_METADATA_URL_TEMPLATE.format(metadata["submission_accession"]))
```
[the remark this answers is not in the document]
Comment (kurtwheeler) [reply foreman/data_refinery_foreman/surveyor/sra.py]: That key should always be present. If it's not it's an error in the XML and the resulting `ValueError` that gets raised seems appropriate.

File: foreman/data_refinery_foreman/surveyor/sra.py [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+        for child in library:
+            if child.tag == "LIBRARY_LAYOUT":
+                metadata["library_layout"] = child[0].tag
+            else:
+                metadata[child.tag.lower()] = child.text
+
+        # SRA contains several types of data. We only want RNA-Seq for now.
+        if metadata["library_strategy"] != "RNA-Seq":
+            raise UnsupportedDataTypeError("library_strategy not RNA-Seq.")
+        if metadata["library_source"] not in ["TRANSCRIPTOMIC", "OTHER"]:
```
[the remark this answers is not in the document]
Comment (kurtwheeler) [reply foreman/data_refinery_foreman/surveyor/sra.py]: Once again, they should. These values are important later so if they are not present then the resulting `ValueError` is appropriate.

File: foreman/data_refinery_foreman/surveyor/sra.py [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+                    elif grandchild.tag == "SCIENTIFIC_NAME":
+                        metadata["organism_name"] = grandchild.text.upper()
+            elif child.tag == "SAMPLE_ATTRIBUTES":
+                for grandchild in child:
+                    key, value = SraSurveyor.parse_attribute(grandchild, "sample_")
+                    metadata[key] = value
+
+    @staticmethod
+    def gather_study_metadata(metadata: Dict) -> None:
+        metadata = {"study_accession": "DRP000595"}
```
[the remark this answers is not in the document]
Comment (kurtwheeler) [reply foreman/data_refinery_foreman/surveyor/sra.py]: Good catch! This was added for development purposes and should have been removed.

File: foreman/data_refinery_foreman/surveyor/sra.py [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+                       release_date=metadata.pop("run_ena_first_public"),
+                       last_uploaded_date=metadata.pop("run_ena_last_update"),
+                       files=files,
+                       key_values=metadata)
+
+    @staticmethod
+    def get_next_accession(last_accession: str) -> str:
+        """Increments a SRA accession number by one.
+
+        E.g. if last_accession is "DRR002116" then "DRR002117" will be
```
[the remark this answers is not in the document]
Comment (kurtwheeler) [reply foreman/data_refinery_foreman/surveyor/sra.py]: Good point actually, I think the length of digits can be variable. I should change this to format it to the original length.

File: foreman/data_refinery_foreman/surveyor/sra.py [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+        survey_job = SurveyJob.objects.get(id=self.survey_job.id)
+        survey_job_properties = survey_job.get_properties()
+
+        logger.info("Surveying SRA runs with accessions in the range of %s to %s.",
+                    survey_job_properties["start_accession"],
+                    survey_job_properties["end_accession"])
+
+        current_accession = survey_job_properties["start_accession"]
+
+        # By evaluating this conditional at the end of the loop
```
[the remark this answers is not in the document]
Comment (kurtwheeler) [reply foreman/data_refinery_foreman/surveyor/sra.py]: I actually did mean `conditional` here, as in a conditional expression. Do you think this isn't an appropriate usage of that term?

File: foreman/data_refinery_foreman/surveyor/sra.py [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+        # instead of the beginning, we achieve the functionality of a
+        # do-while loop.
+        surveyed_last_accession = False
```
[the remark this answers is not in the document]
Comment (kurtwheeler) [reply foreman/data_refinery_foreman/surveyor/sra.py]: .... I somehow forgot to make this into a loop. Clearly I need to add a unit test to this.

File: foreman/data_refinery_foreman/surveyor/test_sra_xml.py (hunk 1 above)
Comment (kurtwheeler) [reply foreman/data_refinery_foreman/surveyor/test_sra_xml.py]: You don't need to check it at all, except as far as it is useful for understanding the XML parsing.

File: foreman/data_refinery_foreman/surveyor/sra.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (dongbohu) [reply foreman/data_refinery_foreman/surveyor/sra.py]: Isn't `conditional` an adjective? English is not my native language, so if you think this usage is okay here, I trust you.

File: common/data_refinery_common/models/surveys.py [hunk 9]
```diff
@@ -20,6 +21,9 @@ class SurveyJob(TimeTrackedModel):
     # The end time of the job
     end_time = models.DateTimeField(null=True)
 
+    def get_properties(self) -> Dict:
+        return {pair.key: pair.value for pair in self.surveyjobkeyvalue_set.all()}
```
[the remark this answers is not in the document]
Comment (kurtwheeler) [reply common/data_refinery_common/models/surveys.py]: It is built into Django models for [`many-to-one` relationships.](https://docs.djangoproject.com/en/1.11/topics/db/examples/many_to_one/)

File: foreman/data_refinery_foreman/surveyor/management/commands/survey_sra.py [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+logger = get_and_configure_logger(__name__)
+
+
+class Command(BaseCommand):
+    def add_arguments(self, parser):
+        parser.add_argument(
+            "start_accession",
+            help=("An SRA run accession. Runs going from the start_accession to the "
+                  "end_accession will be surveyed. If a surveying a single run accession "
```
[the remark this answers is not in the document]
Comment (kurtwheeler) [reply foreman/data_refinery_foreman/surveyor/management/commands/survey_sra.py]: Absolutely, good catch!

File: foreman/data_refinery_foreman/surveyor/management/commands/survey_sra.py [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+                  "is desired then the same accession should be passed to start_accession "
+                  "and end_accession."))
+
+    def handle(self, *args, **options):
+        if options["start_accession"] is None or options["start_accession"] is None:
+            logger.error("You must specify start and end run accessions.")
+            return 1
+        else:
+            surveyor.survey_sra_experiments(options["start_accession"], options["end_accession"])
+            return 0
```
[the remark this answers is not in the document]
Comment (kurtwheeler) [reply foreman/data_refinery_foreman/surveyor/management/commands/survey_sra.py]: I am not currently, but if I ever wanted to call this from a bash script they may prove useful.

File: foreman/data_refinery_foreman/surveyor/management/commands/survey_sra.py [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+            "start_accession",
+            help=("An SRA run accession. Runs going from the start_accession to the "
+                  "end_accession will be surveyed. If a surveying a single run accession "
+                  "is desired then the same accession should be passed to start_accession "
+                  "and end_accession."))
+        parser.add_argument(
+            "end_accession",
+            help=("An SRA run accession. Runs going from the start_accession to the "
+                  "end_accession will be surveyed. If a surveying a single run accession "
+                  "is desired then the same accession should be passed to start_accession "
```
[the remark this answers is not in the document]
Comment (kurtwheeler) [reply foreman/data_refinery_foreman/surveyor/management/commands/survey_sra.py]: Am I correct that this comment is redundant with https://github.com/data-refinery/data_refinery/pull/70#discussion_r147444920?

[the remark this answers is not in the document]
Comment (dongbohu) [reply foreman/data_refinery_foreman/surveyor/management/commands/survey_sra.py]: Yes. I have no idea why it showed up twice.

File: foreman/data_refinery_foreman/surveyor/sra.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (kurtwheeler) [reply foreman/data_refinery_foreman/surveyor/sra.py]: The term has a specific meaning within [computer science.](https://en.wikipedia.org/wiki/Conditional_(computer_programming))

[the remark this answers is not in the document]
Comment (dongbohu) [reply foreman/data_refinery_foreman/surveyor/sra.py]: Good to know. I never took CS101 in English. Forgive me for my ignorance.

[the remark this answers is not in the document]
Comment (kurtwheeler) [reply foreman/data_refinery_foreman/surveyor/sra.py]: I don't believe ignorance is something that should be apologized for :grinning:

**Comment by kurtwheeler** [comment]
@dongbohu I believe I have remedied or responded to all of your comments. Take a look and let me know what you think. I added a few comments to the XML parsing, but mostly just one big comment in the class' docstring. I tried to avoid adding comments to anything that was just doing simple XML parsing.

File: common/data_refinery_common/models/surveys.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (dongbohu) [reply common/data_refinery_common/models/surveys.py]: Thanks.

**Pull request merged**